src/rtc.rs

```diff
@@ -0,0 +1,723 @@
+            11 => self.reg.bkp11r.write(|w| unsafe { w.bkp().bits(value) }),
+            12 => self.reg.bkp12r.write(|w| unsafe { w.bkp().bits(value) }),
+            13 => self.reg.bkp13r.write(|w| unsafe { w.bkp().bits(value) }),
+            14 => self.reg.bkp14r.write(|w| unsafe { w.bkp().bits(value) }),
+            15 => self.reg.bkp15r.write(|w| unsafe { w.bkp().bits(value) }),
+            16 => self.reg.bkp16r.write(|w| unsafe { w.bkp().bits(value) }),
+            17 => self.reg.bkp17r.write(|w| unsafe { w.bkp().bits(value) }),
+            18 => self.reg.bkp18r.write(|w| unsafe { w.bkp().bits(value) }),
+            19 => self.reg.bkp19r.write(|w| unsafe { w.bkp().bits(value) }),
+            20 => self.reg.bkp20r.write(|w| unsafe { w.bkp().bits(value) }),
+            21 => self.reg.bkp21r.write(|w| unsafe { w.bkp().bits(value) }),
+            22 => self.reg.bkp22r.write(|w| unsafe { w.bkp().bits(value) }),
+            23 => self.reg.bkp23r.write(|w| unsafe { w.bkp().bits(value) }),
+            24 => self.reg.bkp24r.write(|w| unsafe { w.bkp().bits(value) }),
+            25 => self.reg.bkp25r.write(|w| unsafe { w.bkp().bits(value) }),
+            26 => self.reg.bkp26r.write(|w| unsafe { w.bkp().bits(value) }),
+            27 => self.reg.bkp27r.write(|w| unsafe { w.bkp().bits(value) }),
+            28 => self.reg.bkp28r.write(|w| unsafe { w.bkp().bits(value) }),
+            29 => self.reg.bkp29r.write(|w| unsafe { w.bkp().bits(value) }),
+            30 => self.reg.bkp30r.write(|w| unsafe { w.bkp().bits(value) }),
+            31 => self.reg.bkp31r.write(|w| unsafe { w.bkp().bits(value) }),
+            _ => panic!("Backup reg index out of range"),
+        }
+
+        // TODO: stm32h7 0.13.0
+        // self.reg.bkpr[reg as usize].write(|w| w.bkp().bits(value));
+    }
+
+    /// Sets the date and time of the RTC
+    pub fn set_date_time(&mut self, date_time: NaiveDateTime) {
+        // Enter initialization mode
+        self.reg.isr.modify(|_, w| w.init().set_bit());
+        while self.reg.isr.read().initf().bit_is_clear() {}
+
+        let hour = date_time.hour() as u8;
+        let ht = hour / 10;
+        let hu = hour % 10;
+
+        let minute = date_time.minute() as u8;
+        let mnt = minute / 10;
+        let mnu = minute % 10;
+
+        let second = date_time.second() as u8;
+        let st = second / 10;
+        let su = second % 10;
+
+        self.reg.tr.write(|w| unsafe {
+            w.pm()
+                .clear_bit()
+                .ht()
+                .bits(ht)
+                .hu()
+                .bits(hu)
+                .mnt()
+                .bits(mnt)
+                .mnu()
+                .bits(mnu)
+                .st()
+                .bits(st)
+                .su()
+                .bits(su)
+        });
+
+        let year = date_time.year();
+        let yt = ((year - 2000) / 10) as u8;
+        let yu = ((year - 2000) % 10) as u8;
```

Comment:
What happens if `year` = 2000? I think `calendar_initialized` returns `None`, because the year field in the register is 0.

What happens if `year` < 2000 or `year` > 2099 ? I think we should handle those cases. I'd say `let year = year % 100;` is OK, although then value read back by the `date()` method will be different. Maybe combine that with a `debug_assert`?

Reply:
The RTC date register only supports years 2000-2099. I'll add an assert there, though good catch.

Reply:
Ok, great. I still think `calendar_initialized` returns None for `year` = 2000, which is the case in the example.

Reply:
Right, the `INITS` just compares the year value to 0 to see if the calendar is initialized. So let's not allow the year 2000.